eng/helix/content/RunTests/Program.cs

```diff
@@ -227,9 +288,57 @@ static async Task Main(string[] args)
             {
                 Console.WriteLine("No logs found in artifacts/log");
             }
+        }
+
+    }
 
-            Console.WriteLine("Completed Helix job.");
-            Environment.Exit(exitCode);
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            try 
+            {
+                var runner = new TestRunner(RunTestsOptions.Parse(args));
+
+                var keepGoing = runner.SetupEnvironment();
+                if (keepGoing) {
+                    keepGoing = await runner.InstallAspNetAppIfNeededAsync();
+                }
+
+                Console.WriteLine();
+                Console.WriteLine("Displaying directory contents:");
+                foreach (var file in Directory.EnumerateFiles("./"))
+                {
+                    Console.WriteLine(Path.GetFileName(file));
+                }
+                foreach (var file in Directory.EnumerateDirectories("./"))
+                {
+                    Console.WriteLine(Path.GetFileName(file));
+                }
+                Console.WriteLine();
+
+                if (keepGoing)
+                {
+                    if (!await runner.CheckTestDiscoveryAsync()) {
+                        Console.WriteLine("RunTest stopping due to test discovery failure.");
+                        Environment.Exit(1);
+                        return;
+                    }
+
+                    var exitCode = await runner.RunTestsAsync();
+                    runner.UploadResults();
+                    Console.WriteLine($"Completed Helix job with exit code '{exitCode}'");
+                    Environment.Exit(exitCode);
+                }
+
+                Console.WriteLine("Tests were not run due to previous failures. Exit code=1");
+                Environment.Exit(1);
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"RunTests uncaught exception: {e.ToString()}");
```

Comment:
Does runtests do any background processing? SHould we listed for background unhandled exceptions?

Reply:
Nothing in the background. It starts processes inline and awaits the TCS which is set by process exit.